qdrant-landing/content/articles/indexing-optimization.md

```diff
@@ -0,0 +1,188 @@
+
+**`indexing_threshold=0`** 
+ 
+Disables HNSW index creation for dense vectors. Qdrant will not build the HNSW graph for those vectors, letting you upload large volumes of data without incurring the memory cost of index creation.
+
+**`indexing_threshold>0`**  
+
+A positive threshold tells Qdrant how many unindexed dense vectors can accumulate in a segment before building the HNSW graph. Small thresholds (e.g., 100) mean more frequent indexing with less data each time, which can still be costly if many segments exist. Larger thresholds (e.g., 10000) delay indexing to batch more vectors at once, potentially using more RAM at the moment of index build, but fewer builds overall. 
+
+```json
+PATCH /collections/your_collection
+{
+  "optimizer_config": {
+    "indexing_threshold": 10000
+  }
+}
+```
+
+<aside role="status">Sparse vectors are always indexed on upsert, regardless of the threshold.</aside>
+
+---
+
+### The `"m"` parameter
+
+For dense vectors, the `m` parameter defines how many edges each node in the HNSW graph can have. Setting `"m": 0` effectively **disables the HNSW graph**, so no dense vector index will be built, no matter the `indexing_threshold`. This can be helpful during massive ingestion if you don’t need immediate searchability.
+
+---
+
+## On-Disk Indexing in Qdrant
+
+By default, Qdrant keeps **vectors and indexes** in **memory** to ensure low-latency queries. However, in large-scale or memory-constrained scenarios, you can configure some or all of those indexes to be stored on-disk. This helps reduce RAM usage at the cost of potential increases in query latency, particularly for cold reads.
+
+**When to use on-disk indexing**:
+- You have **very large** or **rarely used** payload indexes, and freeing up RAM is worth potential I/O overhead.  
+- Your dataset doesn’t fit comfortably in available memory.  
+- You can tolerate slower queries if it ensures the system remains stable under heavy loads.
+
+---
+
+
+## Memmap storage and segmentation
+
+Qdrant uses **memory-mapped files** (segments) to store data on-disk. Rather than loading all vectors into RAM, Qdrant maps each segment into its address space, paging data in and out on demand. This helps keep the active RAM footprint lower, but each segment still incurs overhead (metadata, page table entries, etc.).
+
+During **high-volume ingestion**, you can accumulate dozens of small segments. Qdrant’s **optimizer** can later merge these into fewer, larger segments, reducing per-segment overhead and lowering total memory usage.
+
+When you create a collection with `"on_disk": true`, Qdrant will store newly inserted vectors in memmap storage from the start. For example:
+
+```json
+PATCH /collections/your_collection
+{
+    "vectors": {
+      "on_disk": true
+    }
+}
+```
+
+This approach immediately places all incoming vectors on disk. If you want both dense and sparse vectors to be stored on disk, you need to enable `on_disk` for each type separately.
```

Comment:
```suggestion
This approach immediately places all incoming vectors on disk, which can be very efficient in case of bulk ingestion. If you want both dense and sparse vectors to be stored on disk, you need to enable `on_disk` for each type separately.
```